README.md
# iptables Scripts

iptables Scripts provides a set of scripts of easily managaing and configuring
your iptables and ip6tables based firewalls on your unix-based setup.

## Original work
The scripts are originally provided by [Ray-works.de]. So kudos to him. Because
I am using those scripts a lot and made some modifications to them, I decided to
put them under MIT license and publish them on GitHub. This way, people can
easier find and use them.

## Download
Just download and extract the latest release of the scripts to your server:
```
$ wget https://github.com/moritzrupp/iptables-scripts/releases/latest
$ tar -zxvf iptables-scripts-{version}.tar.gz
```

I recommend to copy the two scripts to `/usr/local/bin` so that they are
available in the path.
```
$ cp iptables-scripts/* /usr/local/bin
```

## Configuration
Now you are ready to configure your firewall easily. Just edit the files using
your preferred text editor. I'm preferring `vi`, but you can use whichever
editor you want:
```
$ vi /usr/local/bin/ipv4settings

$ vi /usr/local/bin/ipv6settings
```

### ipv4settings
Starting from line `42` you can configure the script.

| Parameter   | Values       | Default | Description |
|-------------|--------------|---------|-------------|
| NFS Storage | `yes` / `no` | `no`    | Allow `tcp` and `udp` traffic between the server and NFS storage |
| VPN Forward | `yes` / `no` | `no`    | If you've configured a `VPN` server and want to forward all traffic to the `VPN`, you have to set this to `yes` and configure the `VPN` subnet |
| Fail2Ban    | `yes` / `no` | `no`    | If you use fail2ban, setting this to `yes` will start fail2ban, so that fail2ban rules are re-added |
| Docker      | `yes` / `no` | `no`    | If you have docker installed, you should set this to `yes`. Docker is restarted and traffic outgoing traffic from the server to the docker bridge is allowed. Keep in mind configuring the correct docker subnet |

From line `67` to line `75` you have to configure the ports you want to open.
For example, if you want to open port 22 for ssh, and you want to run a
webserver with http and https:

```
# TCP & UDP Ports for incoming traffic
INTCPPORTS="ssh http https"
INUDPPORTS=""

# TCP & UDP Ports for outgoing traffic
OUTTCPPORTS="ssh http https"
OUTUDPPORTS=""

# SSH Port for extra protection via limits
SSHPORT="22"
```

It is important that you specify the `in tcp ports` and the `out tcp ports`, as
we have set the `OUTPUT` policy to `DROP`.

### ipv6settings
For the `ipv6settings` script, you can only configure your ports starting from
line `33`. It's playing by the same rules as `ipv4settings`.

## Usage
After the configuration of the scripts, it's time to enable your firewall and
proctecting your system.
```
$ ipv4settings starting
Firewall (iptables): enabled.
```

And if you want to allow all access again:
```
$ ipv4settings stop
Firewall (iptables): disabled. (allowing all access)
```

## Remarks
Using this script, your `iptables` configuration is not peristent. After a
reboot, you have to `start` `ipv4settings` and/or `ipv6settings` again. For
Debian, the package [iptables-persistent] can help you.

## Issues
If you have a question, problem or you found a bug, please don't hesitate to
create a [new issue] and get in touch with me.

## Contribution
If you find these scripts useful and you have a use case which is not dealt
with, feel free to contribute to the work. Read more about [contributing].

## License
This work is licensed under the [MIT license].

[Ray-works.de]: https://ray-works.de/
[iptables-persistent]: https://packages.debian.org/search?keywords=iptables-persistent
[MIT license]: https://github.com/moritzrupp/iptables-scripts/blob/master/LICENSE
[new issue]: https://github.com/moritzrupp/iptables-scripts/issues/new
[contributing]: https://github.com/moritzrupp/iptables-scripts/blob/master/CONTRIBUTING.md


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This
project adheres to [Semantic Versioning].

## [Unreleased]

## [1.0.1] - 2016-11-17
### Added
- Support for docker. Docker interface is fetched automatically and a rule is
added for allowing outgoing traffic to the docker bridge.
- Read me file
- Contributing file
- Change log
- License file

### Changed
- The scripts are renamed from `ipv4settings.sh` and `ipv6settings.sh` to
`ipv4settings` and `ipv6settings` and are moved to the `src` folder
- The work is licensed under MIT license

### Removed

## 1.0.0 - 2013-10-29
Last change by [Ray-works.de]. Fully functional scripts with support of VPN
forwarding and NFS storage for `IPv4`. Basic setup for `IPv6`

[Semantic Versioning]: http://semver.org/
[Unreleased]: https://github.com/moritzrupp/iptables-scripts/compare/v1.0.1...HEAD
[1.0.1]: https://github.com/moritzrupp/iptables-scripts/compare/v1.0.0...v1.0.1
[Ray-works.de]: https://ray-works.de


ipv6settings.sh
#!/bin/bash

# ip6tables
# by Ray-works.de
# Update: 21.10.2013

#########################
# Configuration
#########################

INTERFACE=eth0
IPTABLES=`which ip6tables`

# TCP & UDP Ports for incoming traffic
INTCPPORTS=""
INUDPPORTS=""

# TCP & UDP Ports for outgoing traffic
OUTTCPPORTS=""
OUTUDPPORTS=""

# SSH Port for extra protection via limits
SSHPORT=""

#########################
# Turn iptables (for IPv6) on
#########################
function on {
       	
# Flush & default
$IPTABLES -F

# Block Everything
$IPTABLES -P INPUT DROP
$IPTABLES -P OUTPUT DROP
$IPTABLES -P FORWARD DROP

# New Chain for logging
$IPTABLES -N LOGNDROP
$IPTABLES -A LOGNDROP -j LOG -m limit --limit 1/min --log-prefix "[Dropped IPv6]: " --log-level 7
$IPTABLES -A LOGNDROP -j DROP

# New Chain for portscan logging
$IPTABLES -N PORTSCAN
$IPTABLES -A PORTSCAN -j LOG -m limit --limit 1/min --log-prefix "[Portscan IPv6]: " --log-level 7
$IPTABLES -A PORTSCAN -j DROP

# Allow Protocol 41
$IPTABLES -I INPUT 1 -p 41 -j ACCEPT

# Allow established and related connection
$IPTABLES -A INPUT -p tcp -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPTABLES -A OUTPUT -p tcp -m state --state ESTABLISHED,RELATED -j ACCEPT

$IPTABLES -A INPUT -p udp -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPTABLES -A OUTPUT -p udp -m state --state ESTABLISHED,RELATED -j ACCEPT

# Allow ICMP
$IPTABLES -A INPUT -p icmpv6 -j ACCEPT
$IPTABLES -A OUTPUT -p icmpv6 -j ACCEPT

# Allow internal addresses
$IPTABLES -A INPUT -s fe80::/10 --jump ACCEPT
$IPTABLES -A OUTPUT -s fe80::/10 --jump ACCEPT

# Allow DNS Lookup
$IPTABLES -A OUTPUT -p udp --dport 53 -j ACCEPT

#
# TCP & UDP Ports for incoming traffic
#

for PORT in $INTCPPORTS; do
	$IPTABLES -A INPUT -p tcp -i $INTERFACE --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

for PORT in $INUDPPORTS; do
	$IPTABLES -A INPUT -p udp -i $INTERFACE --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

#
# TCP & UDP Ports for outgoing traffic
#

for PORT in $OUTTCPPORTS; do
	$IPTABLES -A OUTPUT -p tcp -o $INTERFACE --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

for PORT in $OUTUDPPORTS; do
	$IPTABLES -A OUTPUT -p udp -o $INTERFACE --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

# Deny more than 3 connection attempts per 10 minutes (SSH)
$IPTABLES -A INPUT -p tcp --dport $SSHPORT -m state --state NEW -m recent --set --name SSH
$IPTABLES -A INPUT -p tcp --dport $SSHPORT -m state --state NEW -m recent --update --seconds 600 --hitcount 4 --rttl --name SSH -j LOGNDROP

# Limit connections per minute from single ip to 10 (HTTP)
$IPTABLES -A INPUT -p tcp --dport 80 -m state --state NEW -m recent --set --name http
$IPTABLES -A INPUT -p tcp --dport 80 -m state --state NEW -m recent --update --seconds 60 --hitcount 10 --rttl --name http -j LOGNDROP

# Rate limit ICMP (ping) packets
$IPTABLES -I INPUT -p ipv6-icmp --icmpv6-type echo-request -m recent --set
$IPTABLES -I INPUT -p ipv6-icmp --icmpv6-type echo-request -m recent --update --seconds 10 --hitcount 5 -j LOGNDROP

# Drop all invalid packets
$IPTABLES -A INPUT -m state --state INVALID -j LOGNDROP
$IPTABLES -A FORWARD -m state --state INVALID -j LOGNDROP
$IPTABLES -A OUTPUT -m state --state INVALID -j LOGNDROP

# Drop new connections without the SYN flag set
$IPTABLES -A INPUT -p tcp ! --syn -m state --state NEW -j PORTSCAN

# syn flood limitation
$IPTABLES -A INPUT -p tcp --syn -m limit --limit 5/s --limit-burst 10 -j LOG --log-prefix "SYN flood: " 
$IPTABLES -A INPUT -p tcp --syn -j DROP

# Portscan: Drop ALL
$IPTABLES -A INPUT -p tcp --tcp-flags ALL ALL -j PORTSCAN

# Portscan: Drop FIN + URG + PSH
$IPTABLES -A INPUT -p tcp --tcp-flags ALL FIN,URG,PSH -j PORTSCAN

# Portscan: Drop nmap Null scan
$IPTABLES -A INPUT -p tcp --tcp-flags ALL NONE -j PORTSCAN

# Portscan: Drop nmap FIN stealth scan
$IPTABLES -A INPUT -p tcp --tcp-flags ALL FIN -j PORTSCAN

# Portscan: Drop XMAS
$IPTABLES -A INPUT -p tcp --tcp-flags ALL URG,ACK,PSH,RST,SYN,FIN -j PORTSCAN

# Portscan: Other combinations
$IPTABLES -A INPUT -p tcp --tcp-flags ALL ACK,RST,SYN,FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags SYN,FIN SYN,FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags SYN,RST SYN,RST -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ALL SYN,RST,ACK,FIN,URG -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags FIN,RST FIN,RST -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,FIN FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,PSH PSH -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,URG URG -j PORTSCAN

echo "Firewall ($IPTABLES): enabled."

}


#########################
# Turn ip6tables off
#########################

function off {

$IPTABLES -F
$IPTABLES -X
$IPTABLES -P INPUT ACCEPT
$IPTABLES -P OUTPUT ACCEPT
$IPTABLES -P FORWARD ACCEPT

echo "Firewall ($IPTABLES): disabled. (allowing all access)"

}

#########################
# Script usage
#########################

case "$1" in
    start)
	on
    ;;
    stop)
	off
    ;;
    restart)
       off
	sleep 3;
       on
    ;;
    *)
	echo "$0 {start|stop|restart}"
	echo "Start executes primary ruleset."
	echo "Stop disables all filtering"
	echo "restart clears then enables"
    ;;
esac

ipv4settings.sh
#!/bin/bash

# iptables Script
# by Ray-works.de
# Update: 29.10.2013

#########################
# Variables
#########################

IPTABLES=`which iptables`

# Interfaces
ETH=eth0
VPN=tun0

# IP Adresses
IP=`ifconfig $ETH | grep inet | cut -d : -f 2 | cut -d \  -f 1`
VPNIP=`ifconfig $VPN | grep inet | cut -d : -f 2 | cut -d \  -f 1`

#########################
# Configuration
#########################

# Allow access to the NFS storage?
NFSSTORAGE="no"

# NFS Storage
NFSIP=""

# Do you wanna use VPN Traffic Forwarding?
VPNFORWARD="no"

# VPN Subnetz
VPNSUB=""

# Do you have/use fail2ban?
FAIL2BAN="no"

# TCP & UDP Ports for incoming traffic
INTCPPORTS=""
INUDPPORTS=""

# TCP & UDP Ports for outgoing traffic
OUTTCPPORTS=""
OUTUDPPORTS=""

# SSH Port for extra protection via limits
SSHPORT=""

# Activate syn cookies (ddos protection).
# Default: 0
echo 1 > /proc/sys/net/ipv4/tcp_syncookies

# Prevents to be a part of an DDOS attack (smurf).
echo 1 > /proc/sys/net/ipv4/icmp_echo_ignore_broadcasts 

# Required
echo 1 > /proc/sys/net/ipv4/ip_forward

#########################
# Turn iptables (for IPv4) on
#########################
function on {
       	
# Flush & default
$IPTABLES -F

# Block Everything
$IPTABLES -P INPUT DROP
$IPTABLES -P OUTPUT DROP
$IPTABLES -P FORWARD DROP

# New Chain for logging
$IPTABLES -N LOGNDROP
$IPTABLES -A LOGNDROP -j LOG -m limit --limit 1/min --log-prefix "[Dropped IPv4]: " --log-level 7
$IPTABLES -A LOGNDROP -j DROP

# New Chain for portscan logging
$IPTABLES -N PORTSCAN
$IPTABLES -A PORTSCAN -j LOG -m limit --limit 1/min --log-prefix "[Portscan IPv4]: " --log-level 7
$IPTABLES -A PORTSCAN -j DROP

# Allow Protocol 4
$IPTABLES -I INPUT 1 -p 4 -j ACCEPT

# Allow internal addresses
$IPTABLES -A INPUT -i lo -j ACCEPT
$IPTABLES -A OUTPUT -o lo -j ACCEPT

# Storage Access
if [ "$NFSSTORAGE" = "yes" ]; then
	$IPTABLES -A INPUT -i $ETH -p tcp -s $NFSIP -j ACCEPT
	$IPTABLES -A INPUT -i $ETH -p udp -s $NFSIP -j ACCEPT

	$IPTABLES -A OUTPUT -o $ETH -p tcp -d $NFSIP -j ACCEPT
	$IPTABLES -A OUTPUT -o $ETH -p udp -d $NFSIP -j ACCEPT
fi

# VPN Traffic Forwarding
if [ "$VPNFORWARD" = "yes" ]; then 
	$IPTABLES -A FORWARD -m state --state RELATED,ESTABLISHED -j ACCEPT
	$IPTABLES -A FORWARD -s $VPNSUB -j ACCEPT
	$IPTABLES -A FORWARD -j REJECT
	$IPTABLES -t nat -A POSTROUTING -s $VPNSUB -o $ETH -j MASQUERADE
fi

# Allow established and related connection
$IPTABLES -A INPUT -p tcp -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPTABLES -A OUTPUT -p tcp -m state --state ESTABLISHED,RELATED -j ACCEPT

$IPTABLES -A INPUT -p udp -m state --state ESTABLISHED,RELATED -j ACCEPT
$IPTABLES -A OUTPUT -p udp -m state --state ESTABLISHED,RELATED -j ACCEPT

# Allow ICMP
$IPTABLES -A INPUT -p icmp -j ACCEPT
$IPTABLES -A OUTPUT -p icmp -j ACCEPT

# Allow DNS Lookup
$IPTABLES -A OUTPUT -p udp --dport 53 -j ACCEPT

#
# TCP & UDP Ports for incoming traffic
#

for PORT in $INTCPPORTS; do
	$IPTABLES -A INPUT -p tcp -i $ETH --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

for PORT in $INUDPPORTS; do
	$IPTABLES -A INPUT -p udp -i $ETH --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

#
# TCP & UDP Ports for outgoing traffic
#

for PORT in $OUTTCPPORTS; do
	$IPTABLES -A OUTPUT -p tcp -o $ETH --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

for PORT in $OUTUDPPORTS; do
	$IPTABLES -A OUTPUT -p udp -o $ETH --dport $PORT -m state --state NEW,ESTABLISHED,RELATED -j ACCEPT
done;

# FTPS explicit passive portrange
$IPTABLES -A INPUT -p tcp --dport 40000:40100 -j ACCEPT
$IPTABLES -A INPUT -p udp --dport 40000:40100 -j ACCEPT

# Deny more than 3 connection attempts per 10 minutes (SSH)
$IPTABLES -A INPUT -p tcp --dport $SSHPORT -m state --state NEW -m recent --set --name SSH
$IPTABLES -A INPUT -p tcp --dport $SSHPORT -m state --state NEW -m recent --update --seconds 600 --hitcount 4 --rttl --name SSH -j LOGNDROP

# Limit connections per minute from single ip to 10 (HTTP)
$IPTABLES -A INPUT -p tcp --dport 80 -m state --state NEW -m recent --set --name http
$IPTABLES -A INPUT -p tcp --dport 80 -m state --state NEW -m recent --update --seconds 60 --hitcount 10 --rttl --name http -j LOGNDROP

# Rate limit ICMP (ping) packets
$IPTABLES -I INPUT -p icmp --icmp-type echo-request -m recent --set
$IPTABLES -I INPUT -p icmp --icmp-type echo-request -m recent --update --seconds 20 --hitcount 10 -j LOGNDROP
$IPTABLES -A INPUT -p icmp -m icmp --icmp-type address-mask-request -j LOGNDROP
$IPTABLES -A INPUT -p icmp -m icmp --icmp-type timestamp-request -j LOGNDROP

# Drop all invalid packets
$IPTABLES -A INPUT -m state --state INVALID -j LOGNDROP
$IPTABLES -A FORWARD -m state --state INVALID -j LOGNDROP
$IPTABLES -A OUTPUT -m state --state INVALID -j LOGNDROP

# Drop new connections without the SYN flag set
$IPTABLES -A INPUT -p tcp ! --syn -m state --state NEW -j PORTSCAN

# syn flood limitation
$IPTABLES -A INPUT -p tcp --syn -m limit --limit 5/s --limit-burst 10 -j LOG --log-prefix "SYN flood: " 
$IPTABLES -A INPUT -p tcp --syn -j DROP

# Portscan: Drop ALL
$IPTABLES -A INPUT -p tcp --tcp-flags ALL ALL -j PORTSCAN

# Portscan: Drop FIN + URG + PSH
$IPTABLES -A INPUT -p tcp --tcp-flags ALL FIN,URG,PSH -j PORTSCAN

# Portscan: Drop nmap Null scan
$IPTABLES -A INPUT -p tcp --tcp-flags ALL NONE -j PORTSCAN

# Portscan: Drop nmap FIN stealth scan
$IPTABLES -A INPUT -p tcp --tcp-flags ALL FIN -j PORTSCAN

# Portscan: Drop XMAS
$IPTABLES -A INPUT -p tcp --tcp-flags ALL URG,ACK,PSH,RST,SYN,FIN -j PORTSCAN

# Portscan: Other combinations
$IPTABLES -A INPUT -p tcp --tcp-flags ALL ACK,RST,SYN,FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags SYN,FIN SYN,FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags SYN,RST SYN,RST -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ALL SYN,RST,ACK,FIN,URG -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags FIN,RST FIN,RST -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,FIN FIN -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,PSH PSH -j PORTSCAN
$IPTABLES -A INPUT -p tcp --tcp-flags ACK,URG URG -j PORTSCAN

echo "Firewall ($IPTABLES): enabled."

# Fail2Ban
if [ "$FAIL2BAN" = "yes" ]; then 
	/etc/init.d/fail2ban start
fi

}


#########################
# Turn iptables off
#########################

function off {

$IPTABLES -F
$IPTABLES -t nat -F PREROUTING 
$IPTABLES -t nat -F POSTROUTING
$IPTABLES -X
$IPTABLES -P INPUT ACCEPT
$IPTABLES -P OUTPUT ACCEPT
$IPTABLES -P FORWARD ACCEPT

echo "Firewall ($IPTABLES): disabled. (allowing all access)"

}

#########################
# Script usage
#########################

case "$1" in
    start)
	on
    ;;
    stop)
	off
    ;;
    restart)
       off
	sleep 3;
       on
    ;;
    *)
	echo "$0 {start|stop|restart}"
	echo "Start executes primary ruleset."
	echo "Stop disables all filtering"
	echo "restart clears then enables"
    ;;
esac